### ISSUE TITLE
static files connections not closing

### ISSUE THREAD

**Issue by areaweb2** [pagespeed/ngx_pagespeed#1227]
**ngx_pagespeed** vers. 1.11.33.2-beta
**nginx** vers. 1.10.1

Hello,
I have this problem when enabling PageSpeed, even when bypassing the optimisations with _pagespeed RewriteLevel PassThrough;_

It seems the server doesn't close some connections until it reaches a timeout (30s). The browser waits for these connections to close, so when I open a page it takes forever to download all images.

**Pagespeed on**
![nginx - pagespeed - connections stalling](https://cloud.githubusercontent.com/assets/20515861/16914380/ae30eed4-4cf1-11e6-9590-b027b45315c4.png)

Instead with PageSpeed disabled this doesn't happen.
**Pagespeed Off**
![nginx - pagespeed off - connections no stalling](https://cloud.githubusercontent.com/assets/20515861/16914405/edf77b1e-4cf1-11e6-98ab-a667ef51d71e.png)

**nginx -V**
nginx version: nginx/1.10.1
built by gcc 4.8.4 (Ubuntu 4.8.4-0ubuntu1~12.04.1)
built with OpenSSL 1.0.1 14 Mar 2012
TLS SNI support enabled
configure arguments: 
--add-module=/root/ngx_pagespeed-release-1.11.33.2-beta 
--with-cc=/usr/lib/gcc-mozilla/bin/gcc 
--with-ld-opt=-static-libstdc++ 
--with-cc-opt='-g -O2 -fPIE -fstack-protector --param=ssp-buffer-size=4 -Wformat -Wformat-security -Werror=format-security -D_FORTIFY_SOURCE=2' 
--with-ld-opt='-Wl,-Bsymbolic-functions -fPIE -pie -Wl,-z,relro -Wl,-z,now' 
--prefix=/usr/share/nginx 
--conf-path=/etc/nginx/nginx.conf 
--http-log-path=/var/log/nginx/access.log 
--error-log-path=/var/log/nginx/error.log 
--lock-path=/var/lock/nginx.lock 
--pid-path=/run/nginx.pid 
--http-client-body-temp-path=/var/lib/nginx/body 
--http-fastcgi-temp-path=/var/lib/nginx/fastcgi 
--http-proxy-temp-path=/var/lib/nginx/proxy 
--http-scgi-temp-path=/var/lib/nginx/scgi 
--http-uwsgi-temp-path=/var/lib/nginx/uwsgi 
--with-debug 
--with-pcre-jit 
--with-ipv6 
--with-http_ssl_module 
--with-http_stub_status_module 
--with-http_realip_module 
--with-http_auth_request_module 
--with-http_addition_module 
--with-http_dav_module 
--with-http_gunzip_module 
--with-http_gzip_static_module 
--with-http_v2_module 
--with-http_sub_module 
--with-stream 
--with-stream_ssl_module 
--with-mail 
--with-mail_ssl_module 
--with-threads

**nginx config**

user www-data;
worker_processes 12 ; # 2 * Number of CPUs
pid /var/run/nginx.pid;

events {
	worker_connections  19000;  
	multi_accept on;
}

worker_rlimit_nofile    20000; 

http 
{
    ssi on;
	limit_conn_zone  $binary_remote_addr zone=one:2m;

        sendfile on;
	tcp_nopush on;
	tcp_nodelay on;
	keepalive_timeout 30;
	
    include /etc/nginx/mime.types;
	default_type application/octet-stream;

	##
	# Logging Settings
	##
	access_log /var/log/nginx/access.log;
	error_log /var/log/nginx/error.log;

	##
	# Gzip Settings
	##
	gzip on;
	gzip_disable "msie6";

	gzip_vary on;
	gzip_comp_level 6;
	gzip_min_length  1000;
	gzip_buffers 16 8k;
	gzip_proxied any;
	gzip_http_version 1.1;
	gzip_types text/plain text/css application/json application/x-javascript text/xml application/xml application/xml+rss text/javascript application/javascript text/x-js;	

	##
	# Virtual Host Configs
	##
	include /etc/nginx/conf.d/*.conf;
	include /etc/nginx/sites-enabled/*;

}

**Site .conf**

fastcgi_cache_path /var/www/SITE_it/cache levels=1:2 keys_zone=SITE:1024m;

fastcgi_cache_key "$scheme $request_method $host $request_uri $cookie_cookieOK $cookie_catFavorite"; 

fastcgi_cache_use_stale error timeout invalid_header updating http_500;
fastcgi_ignore_headers "Cache-Control" "Expires" ;
fastcgi_pass_header Set-Cookie;
fastcgi_hide_header "Cache-Control"; 
	
	
	# PAGESPEED ##############################################################
	pagespeed on;

	pagespeed RewriteLevel PassThrough;	
	pagespeed FileCachePath /var/ngx_pagespeed_cache;	
	
	pagespeed StatisticsPath /ngx_pagespeed_statistics;
	pagespeed GlobalStatisticsPath /ngx_pagespeed_global_statistics;
	pagespeed MessagesPath /ngx_pagespeed_message;
	pagespeed ConsolePath /pagespeed_console;
	pagespeed AdminPath /pagespeed_admin;
	pagespeed GlobalAdminPath /pagespeed_global_admin;
	
	# PAGESPEED ##############################################################

server {

    index index.php;

	# PAGESPEED ##############################################################
	
	# Ensure requests for pagespeed optimized resources go to the pagespeed handler
	# and no extraneous headers get set.
	location ~ "\.pagespeed\.([a-z]\.)?[a-z]{2}\.[^.]{10}\.[^.]+" {
	  add_header "" "";
	}
	location ~ "^/pagespeed_static/" { }
	location ~ "^/ngx_pagespeed_beacon$" { }
	# PAGESPEED ##############################################################
	
    set $no_cache 0;

	# POST requests and urls with a query string should always go to PHP
	if ($request_method = POST) {
		set $no_cache 1;
	}   
	
    location / {

		try_files $uri $uri/ /index.php?$args;

		rewrite ^/...
    }    

	location ~ .php$ {
                              ...
	}

	location ~* ^.+\.(js|css|rss|atom|jpg|jpeg|gif|png|ico|xls|bmp)$ {
		access_log off;	
		log_not_found off; 
		expires max;
	}

	location = /favicon.php { access_log off; log_not_found off; }
	location = /robots.txt { access_log off; log_not_found off; }
	location ~ /\. { deny  all; access_log off; log_not_found off; }
}

**Comment by oschaaf** [comment]
It looks like you are using server side includes (SSI), which does not play well with ngx_pagespeed. Does the problem go away if you turn SSI processing off? 

If so, @jeffkaufman posted a workaround here: https://groups.google.com/forum/#!topic/ngx-pagespeed-discuss/IcJCWwXg410

ps - I do not think it is related to the problem at hand, but you also have rate limiting configured. PageSpeed will perform http fetches to get its inputs and should probably not be rate limited (it already restricts itself). Consider setting up LoadFromFile [1] or explitly allowing it to bypsas the rate limiting.

[1] https://developers.google.com/speed/pagespeed/module/domains#ModPagespeedLoadFromFile

**Comment by areaweb2** [comment]
Disabled SSI, doesn't change anything.
Also tried LoadFromFile, still the same result.

Meanwhile I found this in the **nginx log**:


2016/07/18 10:01:08 [info] 9170#9170: [ngx_pagespeed 1.11.33.2-0] No threading detected. Own threads: 1 Rewrite, 1 Expensive Rewrite.
2016/07/18 10:01:08 [info] 9170#9170: pagespeed: rollback gzip, explicit configuration in /etc/nginx/nginx.conf:36
2016/07/18 10:01:16 [alert] 31695#31695: *5511540 open socket #55 left in connection 4
2016/07/18 10:01:16 [alert] 31695#31695: *5511544 open socket #61 left in connection 9
2016/07/18 10:01:16 [alert] 31695#31695: *5511547 open socket #64 left in connection 38
2016/07/18 10:01:16 [alert] 31695#31695: *5511539 open socket #54 left in connection 57
2016/07/18 10:01:16 [alert] 31695#31695: *5511536 open socket #39 left in connection 68
2016/07/18 10:01:16 [alert] 31695#31695: *5511542 open socket #57 left in connection 76
2016/07/18 10:01:16 [alert] 31695#31695: *5511546 open socket #63 left in connection 85
2016/07/18 10:01:16 [alert] 31695#31695: *5511532 open socket #35 left in connection 90
2016/07/18 10:01:16 [alert] 31695#31695: *5511538 open socket #53 left in connection 92
2016/07/18 10:01:16 [alert] 31695#31695: *5511530 open socket #33 left in connection 93
2016/07/18 10:01:16 [alert] 31695#31695: *5511534 open socket #37 left in connection 96
2016/07/18 10:01:16 [alert] 31695#31695: *5511545 open socket #62 left in connection 102
2016/07/18 10:01:16 [alert] 31695#31695: *5511533 open socket #36 left in connection 105
2016/07/18 10:01:16 [alert] 31695#31695: *5511537 open socket #52 left in connection 108
2016/07/18 10:01:16 [alert] 31695#31695: *5511531 open socket #34 left in connection 111
2016/07/18 10:01:16 [alert] 31695#31695: *5511543 open socket #60 left in connection 113
2016/07/18 10:01:16 [alert] 31695#31695: *5511535 open socket #38 left in connection 114
2016/07/18 10:01:16 [alert] 31695#31695: *5511541 open socket #56 left in connection 117
2016/07/18 10:01:16 [alert] 31695#31695: aborting
2016/07/18 10:01:16 [alert] 31696#31696: *5511669 open socket #30 left in connection 3
2016/07/18 10:01:16 [alert] 31696#31696: *5511665 open socket #23 left in connection 10
2016/07/18 10:01:16 [alert] 31696#31696: *5511664 open socket #3 left in connection 32
2016/07/18 10:01:16 [alert] 31696#31696: *5511666 open socket #26 left in connection 40
2016/07/18 10:01:16 [alert] 31696#31696: *5511668 open socket #29 left in connection 44
2016/07/18 10:01:16 [alert] 31696#31696: *5511667 open socket #28 left in connection 73
2016/07/18 10:01:16 [alert] 31696#31696: aborting
2016/07/18 10:01:16 [alert] 31701#31701: *5511557 open socket #13 left in connection 38
2016/07/18 10:01:16 [alert] 31700#31700: *5511456 open socket #68 left in connection 4
2016/07/18 10:01:16 [alert] 31700#31700: *5511471 open socket #83 left in connection 7
2016/07/18 10:01:16 [alert] 31701#31701: aborting
2016/07/18 10:01:16 [alert] 31700#31700: *5511465 open socket #77 left in connection 9
2016/07/18 10:01:16 [alert] 31700#31700: *5511458 open socket #70 left in connection 10
2016/07/18 10:01:16 [alert] 31700#31700: *5511323 open socket #22 left in connection 17
2016/07/18 10:01:16 [alert] 31700#31700: *5511461 open socket #73 left in connection 19
2016/07/18 10:01:16 [alert] 31700#31700: *5511466 open socket #78 left in connection 21
2016/07/18 10:01:16 [alert] 31700#31700: *5511468 open socket #80 left in connection 22
2016/07/18 10:01:16 [alert] 31700#31700: *5511470 open socket #82 left in connection 23
2016/07/18 10:01:16 [alert] 31700#31700: *5511459 open socket #71 left in connection 24
2016/07/18 10:01:16 [alert] 31700#31700: *5511463 open socket #75 left in connection 28
2016/07/18 10:01:16 [alert] 31700#31700: *5511457 open socket #69 left in connection 31
2016/07/18 10:01:16 [alert] 31700#31700: *5511462 open socket #74 left in connection 35
2016/07/18 10:01:16 [alert] 31700#31700: *5511467 open socket #79 left in connection 36
2016/07/18 10:01:16 [alert] 31700#31700: *5511455 open socket #67 left in connection 41
2016/07/18 10:01:16 [alert] 31700#31700: *5511464 open socket #76 left in connection 42
2016/07/18 10:01:16 [alert] 31700#31700: *5511460 open socket #72 left in connection 52
2016/07/18 10:01:16 [alert] 31700#31700: *5511469 open socket #81 left in connection 53
2016/07/18 10:01:16 [alert] 31700#31700: *5511454 open socket #38 left in connection 57
2016/07/18 10:01:16 [alert] 31700#31700: aborting

**Comment by areaweb2** [comment]
Lowering 	_keepalive_timeout_ to 10 seconds makes the connections close earlier , so the wait for free connections is no longer 30s but 10s.
Also I can tell you the resource (image) gets downloaded immediately (it is immediately rendered) but the connection doesn't close for some reason.

**Comment by oschaaf** [comment]
@areaweb2  Could you set nginx's log level to debug, and post the contents of `error.log` for a single page load? (emailing me at oschaaf@we-amp.com is fine too)

**Comment by areaweb2** [comment]
I'm having trouble generating a log only for my request, since it is a high traffic site. Any advice?

**Comment by oschaaf** [comment]
@areaweb2 If you can capture the logs that contain a page load, that would work too. It would be useful if you could append a recognizeable querystring parameter so I can skip to the right part easily by searching. E.g. http://yoursite/?foo=bar

**Comment by hillsp** [comment]
I wonder if this could be related to the error handling problems I found when fixing https://github.com/pagespeed/ngx_pagespeed/issues/674.

**Comment by oschaaf** [comment]
@areaweb2 thanks, I received your logs and had a look. I wonder if you actively make use of `ngx_http_sub_module`? If so, I think that next to SSI we may also not be compatible with that module..

**Comment by areaweb2** [comment]
@oschaaf we probably don't use most of the modules... I copied the configuration from the Ubuntu compile 
 I'll try to recompile without `ngx_http_sub_module` and let you know the results.

**Comment by oschaaf** [comment]
@areaweb2 if you are not actually using `ngx_http_sub_module` then it very probably is not related to the problem, so never mind, sorry. 

Also, I mentioned that SSI may be a problem but after digging up some more information I think that should not be a problem either.

I took another good look at the logs you send me. I may have found a clue: consider this (anonymized) log fragment:

```
2016/07/18 18:11:16 [info] 16304#16304: [ngx_pagespeed 1.11.33.2-0] [0718/181116:INFO:google_message_handler.cc(53)] Serving rewritten resource in-place: http://www.foo.com\
/img/fooLogo1.png
2016/07/18 18:11:16 [debug] 16304#16304: *628438 HTTP/1.1 200 OK
Server: nginx/1.10.1
Content-Type: image/png
Connection: keep-alive
ETag: "5698db86-f53"
Cache-Control: max-age=315360000
Accept-Ranges: bytes
Date: Fri, 15 Jul 2016 22:00:31 GMT
Expires: Fri, 23 Apr 1948 22:18:57 GMT

```

The important part is that there is no `Content-Length` header. This could explain the symptoms.
A simple test on my own machine gives:

```
2016/07/06 08:37:46 [info] 22744#22744: [ngx_pagespeed 1.11.0.0-7962] [0706/083746:INFO:google_message_handler.cc(53)] Serving rewritten resource in-place: http://localhost:8050\
/abckey/test.css
2016/07/06 08:37:46 [debug] 22744#22744: *5 HTTP/1.1 200 OK^M
Content-Type: text/css^M
Connection: keep-alive^M
Server: nginx/1.8.1^M
Last-Modified: Sat, 02 Jul 2016 20:20:35 GMT^M
Vary: Accept-Encoding^M
ETag: "57782213-64"^M
Accept-Ranges: bytes^M
Date: Wed, 06 Jul 2016 06:37:46 GMT^M
Expires: Wed, 06 Jul 2016 06:42:46 GMT^M
Content-Length: 100^M
Cache-Control: max-age=300, s-maxage=10^M
```

This does show a `Content-Length`.

Would it be possible for you to email me the complete nginx configuration? Hopefully I can reproduce the problem using that. I have tried but my attempts so far have been failing.

**Comment by areaweb2** [comment]
@oschaaf interesting find! Indeed after enabling PageSpeed  the files don't have the _Content-length_ header anymore and instead have _Transfer-Encoding: chunked_ . 

I'll try and send you the configuration as soon as possible.

Could these be related ?
[http://serverfault.com/questions/482875/nginx-removes-content-length-header-for-chunked-content](http://serverfault.com/questions/482875/nginx-removes-content-length-header-for-chunked-content)

[http://serverfault.com/questions/529621/forcing-nginx-to-send-content-length-headers-when-serving-static-files-with-gzip](http://serverfault.com/questions/529621/forcing-nginx-to-send-content-length-headers-when-serving-static-files-with-gzip)

**Comment by oschaaf** [comment]
@areaweb2 If responses end up served with a chunked transfer-encoding instead of a content-length header that's OK from a http perspective. Though I wouldn't expect a chunked transfer encoding for static images, it's not something that would explain the issue here.

What I was seeing in the logs though: the header I posted earlier has neither `Transfer-Encoding` nor a `Content-Length` which is not ok: the receiving browser will not be able to tell when it has received all the data in the response body.

I can reproduce serving combinations of `Connection: close` plus `Transfer-Encoding: chunked` headers when turning off keep-alive in nginx (keepalive_timeout 0;) which is a bit weird but also does not explain the problem as far as I can tell. 

I have not succeeded in logging a response similar to what was in your logs yet.

**Comment by areaweb2** [comment]
@oschaaf actually in Chrome I saw the `Transfer-Encoding: chunked` header, I don't know why it's not in the log. 
I've sent the config files via e-mail. 
Thanks

**Comment by oschaaf** [comment]
@areaweb2 got it - thanks!

**Comment by oschaaf** [comment]
@areaweb2 I cannot reproduce this, but this reminds me of https://github.com/pagespeed/ngx_pagespeed/issues/1146

While debugging that issue with the site owner I ended up with two patches, and one of them needed finishing and testing but I never got around to doing so. The problem there showed up when nginx would respin a request to use php to handle special responses (like 404's).

After some testing, I believe the following patch should work, could try it to see if it eliminates your problem?:
https://gist.github.com/oschaaf/4506974ab8a37f28ee0242e4ae933bcf

**Comment by areaweb2** [comment]
Hello,
I’ve applied the patch and I still get files with ‘Transfer-Encoding:chunked’ that don’t close the connection. I’ve attached the patched file, please confirm it is correct.

The interesting thing is that only some files have this problem and between page reloads it’s these same files presenting this problem.
Therefore some files from the same phyisical folder get transferred with Content-Length header, and other with the Transfer-Encoding:chunked header.



Da: Otto van der Schaaf [mailto:notifications@github.com]
Inviato: lunedì 25 luglio 2016 13:39
A: pagespeed/ngx_pagespeed
Cc: AreaWeb - ALYFA; Mention
Oggetto: Re: [pagespeed/ngx_pagespeed] static files connections not closing (#1227)


@areaweb2<https://github.com/areaweb2> I cannot reproduce this, but this reminds me of #1146<https://github.com/pagespeed/ngx_pagespeed/issues/1146>

While debugging that issue with the site owner I ended up with two patches, and one of them needed finishing and testing but I never got around to doing so. The problem there showed up when nginx would respin a request to use php to handle special responses (like 404's).

After some testing, I believe the following patch should work, could try it to see if it eliminates your problem?:
https://gist.github.com/oschaaf/4506974ab8a37f28ee0242e4ae933bcf

**Comment by areaweb2** [comment]
I notice it only happeens on image files (relative path /uploads or /img). CSS files are unaffected.

**Comment by oschaaf** [comment]
After some private discussion, reproduced on the site in a simple flow within ngx_pagespeed.

Somehow there is no `Content-Length` header in the response for this specific url. Curl states `no chunk, no close, no size. Assume close to signal end` and hangs for a while before closing the connection gets closed:

```
curl --silent -vv --raw http://www.foo.it/uploads/2015/03/bar.jpg.pagespeed.ce.foo.jpg > tmp

* Hostname was NOT found in DNS cache
*   Trying 176.58.112.44...
* Connected to www.foo.it (176.58.112.44) port 80 (#0)
> GET /uploads/2015/03/bar.jpg.pagespeed.ce.foo.jpg HTTP/1.1
> User-Agent: curl/7.35.0
> Host: www.foo.it
> Accept: */*
>
< HTTP/1.1 200 OK
* Server nginx/1.10.1 is not blacklisted
< Server: nginx/1.10.1
< Content-Type: image/jpeg
< Connection: keep-alive
< Link: <http://www.foo.it/uploads/2015/03/bar.jpg>; rel="canonical"
< Last-Modified: Sun, 10 May 2015 16:11:34 GMT
< X-Content-Type-Options: nosniff
< Date: Mon, 01 Aug 2016 13:35:33 GMT
< Expires: Mon, 01 Aug 2016 13:40:33 GMT
< Cache-Control: max-age=300,private
< X-Page-Speed: 1.11.33.2-0
* no chunk, no close, no size. Assume close to signal end
<
{ [data not shown]
* Closing connection 0

```

The IPRO version responds like this (no hang, but note the chunked TE instead a CL header):

```
curl --silent -vv --raw http://www.foo.it/uploads/2015/03/bar.jpg > tmp

* Hostname was NOT found in DNS cache
*   Trying 176.58.112.44...
* Connected to www.foo.it (176.58.112.44) port 80 (#0)
> GET /uploads/2015/03/bar.jpg HTTP/1.1
> User-Agent: curl/7.35.0
> Host: www.foo.it
> Accept: */*
>
< HTTP/1.1 200 OK
* Server nginx/1.10.1 is not blacklisted
< Server: nginx/1.10.1
< Content-Type: image/jpeg
< Transfer-Encoding: chunked
< Connection: keep-alive
< Date: Mon, 01 Aug 2016 13:36:40 GMT
< Expires: Mon, 01 Aug 2016 13:41:40 GMT
< Cache-Control: max-age=300
< Last-Modified: Sun, 10 May 2015 16:11:34 GMT
<
{ [data not shown]
* Connection #0 to host www.foo.it left intact
```

The original version (no hang, has a content-length):

```
curl --silent -vv --raw http://www.foo.it/uploads/2015/03/bar.jpg?PageSpeed=off > tmp
* Hostname was NOT found in DNS cache
*   Trying 176.58.112.44...
* Connected to www.foo.it (176.58.112.44) port 80 (#0)
> GET /uploads/2015/03/bar.jpg?PageSpeed=off HTTP/1.1
> User-Agent: curl/7.35.0
> Host: www.foo.it
> Accept: */*
>
< HTTP/1.1 200 OK
* Server nginx/1.10.1 is not blacklisted
< Server: nginx/1.10.1
< Date: Mon, 01 Aug 2016 13:37:20 GMT
< Content-Type: image/jpeg
< Content-Length: 301730
< Last-Modified: Sun, 10 May 2015 16:11:34 GMT
< Connection: keep-alive
< ETag: "554f8336-49aa2"
< Expires: Thu, 31 Dec 2037 23:55:55 GMT
< Cache-Control: max-age=315360000
< Accept-Ranges: bytes
<
{ [data not shown]
* Connection #0 to host www.foo.it left intact
```

**Comment by oschaaf** [comment]
As per private discussion, the system in question runs a 32 bits nginx binary.

So this very probably was fixed earlier by @hillsp via https://github.com/pagespeed/ngx_pagespeed/pull/1196/commits/58e4f13ad06a9cc3718d81e4e730841c11c73e25

To fix your problem, you could either cherry-pick 58e4f13ad06a9cc3718d81e4e730841c11c73e25 or upgrade to the bleeding edge version.

I am closing this issue, but if the problem did not go away after applying one of these options, please re-open!

**Comment by oschaaf** [comment]
this issue was confirmed to be fixed by 58e4f13
